xbmc/filesystem/AddonsDirectory.cpp

```diff
@@ -826,12 +826,20 @@ void CAddonsDirectory::GenerateAddonListing(const CURL& path,
     bool isUpdate = CheckOutdatedOrUpdate(false); // check if it's an available update
     bool hasUpdate = CheckOutdatedOrUpdate(true); // check if it's an outdated addon
 
+    std::string validUpdateVersion;
+    if (hasUpdate)
+    {
+      const auto& mapEntry = addonsWithUpdate.find(addon->ID());
```

Comment:
std::map returns a lightweight iterator, instead of a reference to the underlying object. It doesn't impact performance to drop the `&`. This might also be incorrect, as the iterator might be destructed after this line, and the reference is unsafely dangling at reclaimable memory.

Also, the constness applies to the underlying object, not the iterator, so `const` can be dropped too. The iterator propagates constantness itself without needing to be const.

Reply:
makes sense to do this, i'm wondering if the same goes here
https://github.com/xbmc/xbmc/blob/a816c1759a38410528c53acd208c163fb84c1475/xbmc/filesystem/AddonsDirectory.cpp#L817
and it should better be `std::shared_ptr<IAddon> checkedObject =`

Reply:
Passing by reference is ok with `shared_ptr`.

Reply:
Agreed. The problem is when the object can be dereferenced elsewhere. I actually fixed a crash where a shared_ptr was passed by reference, and the reference count dropped to zero elsewhere. Fixed by passing by value adding another refcount, `std::shared_ptr<IAddon> checkedObject =`. But here reference is fine because it's so short lived, not worth the atomic ref count increment

Reply:
yeah, that sounds good. thanks for the explanation :)